include/simics/dmllib.h

```diff
@@ -1221,6 +1219,454 @@ _deserialize_simple_event_data(
+        domains_attr_list[i] = _serialize_identity(id_infos,
+                                                   elem.data.domains[i]);
+    }
+
+    return SIM_make_attr_list(2, SIM_make_attr_string("after"),
+                              SIM_make_attr_list(4, callback_key_attr,
+                                                 indices_attr, arguments_attr,
+                                                 domains_attr));
+}
+
+UNUSED static set_error_t
+_DML_deserialize_hook_queue(ht_str_table_t *callback_ht,
+                            ht_str_table_t *id_info_ht,
+                            const _id_info_t *id_infos,
+                            attr_value_t val,
+                            _dml_hook_queue_t *q) {
+    uint32 queue_len = SIM_attr_list_size(val);
+    _dml_hook_queue_t temp_q = VNULL;
+    set_error_t error = Sim_Set_Ok;
+    VRESIZE(temp_q, queue_len);
+    uint32 i = 0;
+    for (; i < queue_len; ++i) {
+        attr_value_t item = SIM_attr_list_item(val, i);
+        const char *elem_kind     = SIM_attr_string(SIM_attr_list_item(item,
+                                                                       0));
+        if (unlikely(strcmp(elem_kind, "after") != 0)) {
+            SIM_c_attribute_error("Illegal hook queue elem kind '%s' when "
+                                  "deserializing hook queue", elem_kind);
+            error = Sim_Set_Illegal_Value;
+            goto error;
+        }
+        _dml_hook_queue_elem_t elem;
+        error = _DML_deserialize_hook_queue_elem_after(
+            callback_ht, id_info_ht, id_infos, SIM_attr_list_item(item, 1),
+            &elem);
+        if (unlikely(error != Sim_Set_Ok)) {
+            goto error;
+        }
+        VSET(temp_q, i, elem);
+    }
+    _DML_free_hook_queue(q);
+    *q = temp_q;
+    return Sim_Set_Ok;
+  error:
+    VTRUNCATE(temp_q, i);
+    _DML_free_hook_queue(&temp_q);
+    return error;
+}
+
+
+UNUSED static attr_value_t
+_DML_serialize_hook_queue(ht_str_table_t *callback_ht,
+                          const _id_info_t *id_infos,
+                          _dml_hook_queue_t *q) {
+    attr_value_t out = SIM_alloc_attr_list(VLEN(*q));
+    attr_value_t *out_list = SIM_attr_list(out);
+    VFORI(*q, i) {
+        out_list[i] = _DML_serialize_hook_queue_elem_after(callback_ht,
+                                                           id_infos,
+                                                           VGET(*q, i));
+    }
+    return out;
+}
+
+
+UNUSED static void
+_DML_register_hook_attribute(conf_class_t *cls, const char *attrname,
+                             get_attr_t getter, set_attr_t setter,
+                             const _dml_hook_aux_info_t *aux_info,
+                             _id_info_t hook_id_info, uint32 port_dims) {
+
+    // Conceptually, "[[sa]*]". Since the only elem representation is that of
+    // "after", instead of "a" that representation is used directly.
```

Comment:
note: when expanding this, it will probably make sense to expand the type string to `[[sX]|[sY]|[sZ]*]`, where X,Y,Z are complex expressions, rather than `[[sa]]`.

Reply:
no rather than `[[sa]*]`. If `[[s(X|Y|Z)]*]` is possible then that's even better.